### PR TITLE
[@mantine/form] use-form: Make functions referentially stable

### LINKED ISSUE #1196: Excessive re-rendering from useForm hook
### Discussed in https://github.com/mantinedev/mantine/discussions/1193

<div type='discussions-op-text'>

<sup>Originally posted by **rfdrake27** April  6, 2022</sup>
Hello, I'm noticing that updating any value within the form will trigger a re-render of the parent component where the useForm hook is located. Even after destructuring the returned object from useForm, I'm noticing that functions such as getInputProps or addListItem are also being changed. 

Are there plans to optimize performance of this hook? Such as wrapping functions in useCallback, or using refs?

Thank you Vitaly for the excellent library. </div>

### PR DESCRIPTION
**Pull request by mattrunyon**
Fixes #1196, #1193 

This makes all functions returned by `useForm` stable. I used the eslint react-hooks plugin for their exhaustive dependencies rule while converting these. It helped catch everywhere that the state needed to be switched to using the ref value within the `useCallback` uses

### CODE REVIEW

File: src/mantine-form/src/use-form.ts [hunk 1]
```diff
@@ -63,46 +63,73 @@ export interface UseFormReturnType<T> {
...
 }
 
+/**
+ * Creates a ref that stays up to date with the state of the param
+ * Used to help with all the useCallback signatures in useForm
+ *
+ * @param value The value to clone in the ref
+ * @returns The ref which clones the value
+ */
+function useStateRef<T>(value: T): React.MutableRefObject<T> {
```
Comment (rtivital) [inline src/mantine-form/src/use-form.ts:73]: It should be in a separate file

File: src/mantine-form/src/use-form.ts [hunk 2]
```diff
@@ -63,46 +63,73 @@ export interface UseFormReturnType<T> {
   ) => GetInputProps<L>;
 }
 
+/**
+ * Creates a ref that stays up to date with the state of the param
+ * Used to help with all the useCallback signatures in useForm
+ *
+ * @param value The value to clone in the ref
```
Comment (rtivital) [inline src/mantine-form/src/use-form.ts:70]: No need for documentation in comments, it's pretty clear what hook does

File: src/mantine-form/src/use-form.ts [hunk 3]
```diff
@@ -63,46 +63,73 @@ export interface UseFormReturnType<T> {
...
+}
+
 export function useForm<T extends { [key: string]: any }>({
   initialValues,
   initialErrors,
   validate: rules,
   schema,
 }: UseFormInput<T>): UseFormReturnType<T> {
   const [errors, setErrors] = useState(filterErrors(initialErrors));
+  const errorsRef = useStateRef(errors);
```
Comment (rtivital) [inline src/mantine-form/src/use-form.ts:90]: I think it is too wordy to create separate variables and calling hooks to get refs. Instead move this logic to `useRefState` hook, that will handle that, for example:

```tsx
const [errors, setErrors, errorsRef] = useRefState(filterErrors(initialErrors));
```

**Review by rtivital — COMMENTED** [review]
I haven't checked how it works, but if tests are passing, then it is 👍

**Comment by mattrunyon** [comment]
Made the changes you suggested (and also made another helper hook to track the state of a prop in a ref).

I also added a unit test for checking the functions are stable. Tested it failed when I removed `useCallback` from one of functions returned by the hook

**Comment by rtivital** [comment]
Thanks!

**Comment by rtivital** [comment]
I've merged it without checking how it works in real demos (stupid of me) and it turns out that it does not work at all – all demos in broken (values updates are random). I've reverted the changes, so if you want, you can open another PR for this.
